### PR TITLE
Introduce mir::Unevaluated

### PR DESCRIPTION
**Pull request by b-naber**
Previously the distinction between unevaluated constants in the type-system and in mir was not explicit and a little confusing. Probably better to introduce its own type for that.

r? @lcnr

### CODE REVIEW

File: compiler/rustc_middle/src/mir/mod.rs [hunk 1]
```diff
@@ -2434,6 +2433,31 @@ impl<'tcx> ConstantKind<'tcx> {
     }
 }
 
+/// An unevaluated (potentially generic) constant used in MIR.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd, Ord, TyEncodable, TyDecodable, Lift)]
+#[derive(Hash, HashStable)]
+pub struct Unevaluated<'tcx, P = Option<Promoted>> {
```
Comment (lcnr) [inline compiler/rustc_middle/src/mir/mod.rs:2439]: can this not always be `Option<Promoted>`?

File: compiler/rustc_middle/src/ty/consts/kind.rs [hunk 2]
```diff
@@ -29,23 +28,15 @@ impl rustc_errors::IntoDiagnosticArg for Unevaluated<'_> {
 
 impl<'tcx> Unevaluated<'tcx> {
     #[inline]
-    pub fn shrink(self) -> Unevaluated<'tcx, ()> {
-        debug_assert_eq!(self.promoted, None);
-        Unevaluated { def: self.def, substs: self.substs, promoted: () }
+    pub fn expand(self) -> mir::Unevaluated<'tcx> {
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/consts/kind.rs:31]: i would love for us to not have that method, instead changing shared methods to again use def + substs separately :thinking: afaict that method is currently unused?

**Review by lcnr — COMMENTED** [review]
nice, small nits, then r=me

File: compiler/rustc_middle/src/mir/mod.rs [hunk 3]
```diff
@@ -2434,6 +2433,31 @@ impl<'tcx> ConstantKind<'tcx> {
...
+    pub def: ty::WithOptConstParam<DefId>,
+    pub substs: SubstsRef<'tcx>,
+    pub promoted: P,
+}
+
+impl<'tcx> Unevaluated<'tcx> {
+    // FIXME: probably should get rid of this method. It's also wrong to
+    // shrink and then later expand a promoted.
+    #[inline]
```
Comment (lcnr) [inline compiler/rustc_middle/src/mir/mod.rs:2448]: yeah, once we've removed `visit_mir_constant` we should be able to remove this, probably makes sense to do that in yet another pr :grin:

File: compiler/rustc_middle/src/mir/mod.rs [hunk 4]
```diff
@@ -2440,6 +2439,31 @@ impl<'tcx> ConstantKind<'tcx> {
...
+/// An unevaluated (potentially generic) constant used in MIR.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd, Ord, TyEncodable, TyDecodable, Lift)]
+#[derive(Hash, HashStable)]
+pub struct Unevaluated<'tcx> {
+    pub def: ty::WithOptConstParam<DefId>,
+    pub substs: SubstsRef<'tcx>,
+    pub promoted: Option<Promoted>,
+}
+
+impl<'tcx> Unevaluated<'tcx> {
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/mod.rs:2451]: Calling it just 'unevaluated' is very confusing. `mir::Unevaluated` doesn't show any relationship to constants. It should be `UnevaluatedConst` or so IMO.

Comment (lcnr) [reply compiler/rustc_middle/src/mir/mod.rs:2451]: :shrug: don't have a strong opinion here, `ConstDefId` seems wrong, would be fine with `UnevaluatedConst`

File: compiler/rustc_middle/src/ty/consts/kind.rs (hunk 2 above)
Comment (lcnr) [reply compiler/rustc_middle/src/ty/consts/kind.rs:31]: well, i don't think we should ever convert `ty::Unevaluated` to `mir::Unevaluated` ^^ didn't see that this method is used, going to look at this at a later point then

File: compiler/rustc_const_eval/src/transform/check_consts/qualifs.rs [hunk 5]
```diff
@@ -350,29 +351,21 @@ where
...
         assert!(promoted.is_none() || Q::ALLOW_PROMOTED);
 
         // Don't peek inside trait associated constants.
         if promoted.is_none() && cx.tcx.trait_of_item(def.did).is_none() {
-            let qualifs = if let Some((did, param_did)) = def.as_const_arg() {
-                cx.tcx.at(constant.span).mir_const_qualif_const_arg((did, param_did))
-            } else {
-                cx.tcx.at(constant.span).mir_const_qualif(def.did)
-            };
+            let qualifs = cx.tcx.at(constant.span).mir_const_qualif(def.did);
```
Comment (lcnr) [inline compiler/rustc_const_eval/src/transform/check_consts/qualifs.rs:368]:
```suggestion
            assert_eq!(def.param_did, None, "expected associated const: {def:?});
            let qualifs = cx.tcx.at(constant.span).mir_const_qualif(def.did);
```

**Review by lcnr — COMMENTED** [review]
final nit, then r=me

File: compiler/rustc_mir_transform/src/inline.rs [hunk 6]
```diff
@@ -605,7 +605,7 @@ impl<'tcx> Inliner<'tcx> {
                 caller_body.required_consts.extend(
                     callee_body.required_consts.iter().copied().filter(|&ct| match ct.literal {
                         ConstantKind::Ty(_) => {
-                            bug!("should never encounter ty::Unevaluated in `required_consts`")
+                            bug!("should never encounter ty::UnevaluatedConst in `required_consts`")
```
Comment (lcnr) [inline compiler/rustc_mir_transform/src/inline.rs:608]: this should be "unexpected `ty::Const`"

**Comment by lcnr** [comment]
@bors r+ rollup=never

maybe perf relevant, probably not :shrug: the queue is pretty short rn
